docs/architecture/adr-036-arbitrary-signature.md

```diff
@@ -0,0 +1,103 @@
+# ADR 036: Arbitrary Message Signature Specification
+
+## Changelog
+
+- 28/10/2020 - Initial draft
+
+## Authors
+- Antoine Herzog (@antoineherzog)
+- Zaki Manian (@zmanian)
+- Aleksandr Bezobchuk (alexanderbez) [1]
+- Frojdi Dymylja (@fdymylja)
+
+## Status
+
+Draft
+
+## Abstract
+
+Currently, in the SDK, there is no convention to sign arbitrary message like on Ethereum. We propose with this specification a way to sign arbitrary message for Cosmos SDK chains.
+
+## Context
+
+Having the ability to sign messages off-chain has proven to be a fundamental aspect of nearly any blockchain. The notion of signing messages off-chain has many added benefits such as saving on computational costs and reducing transaction throughput and overhead. Within the context of the Cosmos, some of the major applications of signing such data includes, but is not limited to, providing a cryptographic secure and verifiable means of proving validator identity and possibly associating it with some other framework or organization. In addition, having the ability to sign Cosmos messages with a Ledger or similar HSM device.
+
+
+## Decision
+
+The aim is being able to sign arbitrary messages, even using Ledger or similar HSM devices.
+
+As a result signed messages should look roughly like Cosmos SDK messages. chain-id, account_number and sequence can all be assigned invalid values.
+The CLI should set those to default values.
+
+Cosmos SDK 0.40 also introduces a concept of “auth_info” this can specify SIGN_MODES.
+
+A spec should include an auth info that supports SIGN_MODE_DIRECT and SIGN_MODE_LEGACY_AMINO.
+
+- the memo should be empty
+- nonce, sequence number should be equal to 0
+- chain-id should be equal to “signature”
+- fee gas should be equal to 0
+- fee amount should be an empty array
+- Inside the message with the type MsgSignText, we put inside a  *bytes* message and the address of the signer.
+
+Proto definition:
+```proto
+// MsgSignedMessage defines 
+message MsgSignedMessage {
+    // Signer is the sdk.AccAddress of the message signer
+    bytes Signer = 1 [(gogoproto.jsontag) = "signer", (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    // Message represents the raw bytes of the content that is signed (text, json, etc)
+    bytes Message = 2 [(gogoproto.jsontag) = "message"];
+}
```

Comment:
I would propose the following, based on both the OAuth2 [implicit](https://tools.ietf.org/html/rfc6749#section-4.2.1) and [code grant flows](https://tools.ietf.org/html/rfc6749#section-4.1.1).

```proto
message MsgSignText {
    // we can add other request_types later with different semantics - this is for the obvious "prove your identity" / access control flow
    string request_type = 1;  // MUST be "bauth"
   
    // which client we are identifying with, could be `registry.atlas.io` or `2du6Sje7` but is unique to the site you authenticate with 
    string client_id = 2;     

    // (OPTIONAL) app-dependent if we want to give a partial grant - eg. for a forum, we could auth it to upvote but not make posts
    string scope = 3;

   // (OPTIONAL) a nonce or session token that can prevent replay attacks
   string state = 4;

    // ????? I don't think we need this, but it is in the Oauth2 flow - I belive this is different in the proposed flow.
    // but maybe is useful, I would like an end-to-end client workflow explained.
    string redirect_url = 5;
}
```

Note that the `Signer` would only be needed for actual processing in a module, via `GetSigners()`, but since we don't ever intend this to go to the blockchain, I would remove it as unnecessary data. The pubkey stored in the signature is the information we want to prove. 

This would allow a website to produce an unsigned transaction *without knowing your pubkey* and then you can sign it with any pubkey you control to identify yourself.